README.md
# range-request-middleware

[![deno land](http://img.shields.io/badge/available%20on-deno.land/x-lightgrey.svg?logo=deno)](https://deno.land/x/range_request_middleware)
[![deno doc](https://doc.deno.land/badge.svg)](https://doc.deno.land/https/deno.land/x/range_request_middleware/mod.ts)
[![GitHub release (latest by date)](https://img.shields.io/github/v/release/httpland/range-request-middleware)](https://github.com/httpland/range-request-middleware/releases)
[![codecov](https://codecov.io/github/httpland/range-request-middleware/branch/main/graph/badge.svg)](https://codecov.io/gh/httpland/range-request-middleware)
[![GitHub](https://img.shields.io/github/license/httpland/range-request-middleware)](https://github.com/httpland/range-request-middleware/blob/main/LICENSE)

[![test](https://github.com/httpland/range-request-middleware/actions/workflows/test.yaml/badge.svg)](https://github.com/httpland/range-request-middleware/actions/workflows/test.yaml)
[![NPM](https://nodei.co/npm/@httpland/range-request-middleware.png?mini=true)](https://nodei.co/npm/@httpland/range-request-middleware/)

HTTP range request middleware.

Handles range request and partial response.

Compliant with
[RFC 9110, 14. Range Requests](https://www.rfc-editor.org/rfc/rfc9110#section-14)

## Usage

Upon receipt of a range request, if the response [satisfies](#satisfiable) the
range requirement, [convert](#convert) it to a partial response.

```ts
import { rangeRequest } from "https://deno.land/x/range_request_middleware@$VERSION/middleware.ts";
import {
  assert,
  assertEquals,
  assertThrows,
} from "https://deno.land/std/testing/asserts.ts";

const middleware = rangeRequest();
const request = new Request("test:", {
  headers: { range: "bytes=5-9" },
});
const response = await middleware(
  request,
  () => new Response("abcdefghijklmnopqrstuvwxyz"),
);

assertEquals(response.status, 206);
assertEquals(response.headers.get("content-range"), "bytes 5-9/26");
assertEquals(response.headers.get("accept-ranges"), "bytes");
assertEquals(await response.text(), "fghij");
```

yield:

```http
HTTP/1.1 206
Content-Range: bytes 5-9/26
Accept-Ranges: bytes

fghij
```

## Multi-range request

For multi-range request, response body will convert to a multipart content.

It compliant with
[RFC 9110, 14.6. Media Type multipart/byteranges](https://www.rfc-editor.org/rfc/rfc9110.html#name-media-type-multipart-bytera).

```ts
import { rangeRequest } from "https://deno.land/x/range_request_middleware@$VERSION/middleware.ts";
import {
  assert,
  assertEquals,
  assertThrows,
} from "https://deno.land/std/testing/asserts.ts";

const middleware = rangeRequest();
const request = new Request("test:", {
  headers: { range: "bytes=5-9, 20-, -5" },
});
const response = await middleware(
  request,
  () => new Response("abcdefghijklmnopqrstuvwxyz"),
);

assertEquals(response.status, 206);
assertEquals(
  response.headers.get(
    "content-type",
  ),
  "multipart/byteranges; boundary=<boundary-delimiter>",
);
assertEquals(
  await response.text(),
  `--<boundary-delimiter>
Content-Type: text/plain;charset=UTF-8
Content-Range: 5-9/26

fghij
--<boundary-delimiter>
Content-Type: text/plain;charset=UTF-8
Content-Range: 20-25/26

uvwxyz
--<boundary-delimiter>
Content-Type: text/plain;charset=UTF-8
Content-Range: 21-25/26

vwxyz
--<boundary-delimiter>--`,
);
```

yield:

```http
HTTP/1.1 206
Content-Type: multipart/byteranges; boundary=BOUNDARY
Accept-Ranges: bytes

--BOUNDARY
Content-Type: text/plain;charset=UTF-8
Content-Range: 5-9/26

fghij
--BOUNDARY
Content-Type: text/plain;charset=UTF-8
Content-Range: 20-25/26

uvwxyz
--BOUNDARY
Content-Type: text/plain;charset=UTF-8
Content-Range: 21-25/26

vwxyz
--BOUNDARY--
```

## Conditions

There are several conditions that must be met in order for middleware to
execute.

If the following conditions are **not met**,
[invalid](https://www.rfc-editor.org/rfc/rfc9110#section-14.2-6) and the
response will not [convert](#convert).

- Request method is `GET`.
- Request includes `Range` header
- Request does not include `If-Range` header
- Request `Range` header is valid syntax
- Request `Range` header is valid semantics
- Response status code is `200`
- Response does not include `Content-Range` header
- Response does not include `Accept-Ranges` header or its value is not `none`
- Response body is readable

Note that if there is an `If-Range` header, do nothing.

## Unsatisfiable

If [conditions](#conditions) is met and the following conditions are **not met**
,[unsatisfiable](https://www.rfc-editor.org/rfc/rfc9110#section-14.1.1-12), and
it is not possible to meet partial response.

- If a valid
  [ranges-specifier](https://www.rfc-editor.org/rfc/rfc9110#rule.ranges-specifier)
  contains at least one satisfactory
  [range-spec](https://www.rfc-editor.org/rfc/rfc9110#rule.ranges-specifier), as
  defined in the indicated
  [range-unit](https://www.rfc-editor.org/rfc/rfc9110#range.units)

In this case, the handler response will [convert](#convert) to
[416(Range Not Satisfiable)](https://www.rfc-editor.org/rfc/rfc9110#status.416)
response.

A example of how unsatisfiable can happen:

If receive un unknown range unit.

```ts
import {
  type Handler,
  rangeRequest,
} from "https://deno.land/x/range_request_middleware@$VERSION/mod.ts";
import { assert, assertEquals } from "https://deno.land/std/testing/asserts.ts";

declare const handler: Handler;
const middleware = rangeRequest();
const response = await middleware(
  new Request("test:", { headers: { range: "<unknown-unit>=<other-range>" } }),
  handler,
);

assertEquals(response.status, 416);
assert(response.headers.has("content-range"));
```

## Satisfiable

If the [conditions](#conditions) and [unsatisfiable](#unsatisfiable) are met,
[satisfiable](https://www.rfc-editor.org/rfc/rfc9110#satisfiable), and the
response will [convert](#convert) to
[206(Partial Content)](https://www.rfc-editor.org/rfc/rfc9110#section-15.3.7)
response.

## Convert

Convert means a change without side effects.

For example, "convert a response to the 206 response" means to return a new
response in which some or all of the following elements have been replaced from
the original response.

- HTTP Content
- HTTP Status code
- HTTP Headers(shallow marge)

## Range

`Range` abstracts partial response.

Middleware factories can accept `Range` objects and implement own range request
protocols.

`Range` is the following structure:

| Name      | Type                                                                                      | Description                                 |
| --------- | ----------------------------------------------------------------------------------------- | ------------------------------------------- |
| rangeUnit | `string`                                                                                  | Corresponding range unit.                   |
| respond   | `(response: Response, context: RangesSpecifier) =>` `Response` &#124; `Promise<Response>` | Return response from range request context. |

The middleware supports the following range request protocols by default:

- `bytes`([ByteRanges](#bytesrange))

### BytesRange

`bytes` range unit is used to express subranges of a representation data's octet
sequence.

ByteRange supports single and multiple range requests.

Compliant with
[RFC 9110, 14.1.2. Byte Ranges](https://www.rfc-editor.org/rfc/rfc9110.html#section-14.1.2).

```ts
import {
  BytesRange,
  type IntRange,
  type SuffixRange,
} from "https://deno.land/x/range_request_middleware@$VERSION/mod.ts";
import { assertEquals } from "https://deno.land/std/testing/asserts.ts";

const bytesRange = new BytesRange();
const rangeUnit = "bytes";
declare const initResponse: Response;
declare const rangeSet: [IntRange, SuffixRange];

const response = await bytesRange.respond(initResponse, {
  rangeUnit,
  rangeSet,
});

assertEquals(bytesRange.rangeUnit, rangeUnit);
assertEquals(response.status, 206);
assertEquals(
  response.headers.get("content-type"),
  "multipart/byteranges; boundary=<BOUNDARY>",
);
```

## Effects

Middleware may make changes to the following HTTP messages:

- HTTP Content
- HTTP Headers
  - Accept-Ranges
  - Content-Range
  - Content-Type
- HTTP Status code
  - 206(Partial Content)
  - 416(Range Not Satisfiable)

## License

Copyright © 2023-present [httpland](https://github.com/httpland).

Released under the [MIT](./LICENSE) license


CHANGELOG.md
# [1.3.0](https://github.com/httpland/range-request-middleware/compare/1.2.0...1.3.0) (2023-05-02)


### Features

* **deps:** update deps ([370f1fc](https://github.com/httpland/range-request-middleware/commit/370f1fc84425dcc849c2a844eb395b15f751a7eb))

# [1.3.0-beta.1](https://github.com/httpland/range-request-middleware/compare/1.2.0...1.3.0-beta.1) (2023-04-29)


### Features

* **deps:** update deps ([370f1fc](https://github.com/httpland/range-request-middleware/commit/370f1fc84425dcc849c2a844eb395b15f751a7eb))

# [1.2.0](https://github.com/httpland/range-request-middleware/compare/1.1.0...1.2.0) (2023-04-02)


### Features

* **deps:** update deps version ([de20d9a](https://github.com/httpland/range-request-middleware/commit/de20d9aafda8baebd0a12a4209fc07b516463503))

# [1.2.0-beta.1](https://github.com/httpland/range-request-middleware/compare/1.1.0...1.2.0-beta.1) (2023-04-02)


### Features

* **deps:** update deps version ([de20d9a](https://github.com/httpland/range-request-middleware/commit/de20d9aafda8baebd0a12a4209fc07b516463503))

# [1.1.0](https://github.com/httpland/range-request-middleware/compare/1.0.0...1.1.0) (2023-03-29)


### Features

* **deps:** use external content-range serializer ([5095c21](https://github.com/httpland/range-request-middleware/commit/5095c213075b02d76d1ce03ed37c26511092117e))
* **mod:** export dependent interface ([32609a6](https://github.com/httpland/range-request-middleware/commit/32609a67d0249d100b74e21b66a35a4b71a959b4))

# [1.1.0-beta.2](https://github.com/httpland/range-request-middleware/compare/1.1.0-beta.1...1.1.0-beta.2) (2023-03-29)


### Features

* **deps:** use external content-range serializer ([5095c21](https://github.com/httpland/range-request-middleware/commit/5095c213075b02d76d1ce03ed37c26511092117e))

# [1.1.0-beta.1](https://github.com/httpland/range-request-middleware/compare/1.0.0...1.1.0-beta.1) (2023-03-28)


### Features

* **mod:** export dependent interface ([32609a6](https://github.com/httpland/range-request-middleware/commit/32609a67d0249d100b74e21b66a35a4b71a959b4))

# 1.0.0 (2023-03-26)


### Bug Fixes

* **transform:** change the range unit equality to case insensitive ([16d547e](https://github.com/httpland/range-request-middleware/commit/16d547e067a8c61eeb9400c2d49f01f7b2d440fc))
* **transform:** change the success status code ([7e789a1](https://github.com/httpland/range-request-middleware/commit/7e789a18076f97d1ea1c27480a5bfa205ea9915e))
* **transform:** fix to change 416 response condition ([2509e90](https://github.com/httpland/range-request-middleware/commit/2509e906f0e8f7a0961740a454ea83826c4d1176))
* **transform:** remove unnecessary requirements ([43900d8](https://github.com/httpland/range-request-middleware/commit/43900d8f423c563541cd2f216fa89825f0d04114))


### Features

* **accept_ranges:** add accept-ranges header parser and check that header is none strictly ([980f9b4](https://github.com/httpland/range-request-middleware/commit/980f9b4b8d6b28460db028fd75a4dd9eaff86950))
* add content-range header and partilize content ([165d180](https://github.com/httpland/range-request-middleware/commit/165d180d63ad432cc60210bd39be60e09ed93c87))
* **bytes:** export BytesRange and add description ([253386f](https://github.com/httpland/range-request-middleware/commit/253386ff8c9ddf5ee423aec4319d213876a02817))
* **content_range:** add checking format of content-range header ([169b15d](https://github.com/httpland/range-request-middleware/commit/169b15d8d90fadaca9e8161c41c21e9fed84600c))
* **content_range:** add content range field deserializer ([0fe7759](https://github.com/httpland/range-request-middleware/commit/0fe775969a9117f8a2a55b402b35de9456e88790))
* **middleware:** add range middleware factory ([1926b02](https://github.com/httpland/range-request-middleware/commit/1926b027502c1666661bc3e012eaf8b0e4faf13f))
* **middleware:** add rangeRequest, contentRange and acceptRanges middleware ([03e8d6b](https://github.com/httpland/range-request-middleware/commit/03e8d6b75a003aaa382f282ee916fbec4ab5b100))
* **mod:** export types of interface ([8687894](https://github.com/httpland/range-request-middleware/commit/8687894131197c1c9a2bb95df29fd7cd7c4487cf))
* **mod:** remove unrelated types ([068dd87](https://github.com/httpland/range-request-middleware/commit/068dd87cdf0870eac0e6534a214423b6d61ca54f))
* **parser:** add range header parser ([4d5a9ab](https://github.com/httpland/range-request-middleware/commit/4d5a9ab3494f7d97b91f8f44bc8be7c0e1eea634))
* **ranges:** add ByteRange implementation ([1741e1d](https://github.com/httpland/range-request-middleware/commit/1741e1d26437bcadd2286f18dc30d6e0c54637c5))
* rename field name to `rangeUnit` from `unit` ([41aa7da](https://github.com/httpland/range-request-middleware/commit/41aa7da5170ffd1a789fa9b5135cd8a5a57c8ee3))
* **transform:** add response transformer for range headers ([0f1700c](https://github.com/httpland/range-request-middleware/commit/0f1700c2cd1bac2d5a0a57ee6419640bcd9398a1))
* **transform:** add to transformer for `accept-ranges` header field ([5f7f226](https://github.com/httpland/range-request-middleware/commit/5f7f2260efde91c7b687dc1938d9c085ecb8eff2))
* **transform:** change Range interface and transform response interface ([ff78387](https://github.com/httpland/range-request-middleware/commit/ff783875603d87702aa69393b31631f71204aacf))
* **types:** add Range API ([50d9eab](https://github.com/httpland/range-request-middleware/commit/50d9eabd343e0108527414e08f7e0565ccee2980))
* **types:** change range interface, update docs ([28b2da9](https://github.com/httpland/range-request-middleware/commit/28b2da91541b60365db43ebdaa7a120ff607cdac))
* **types:** remove object method types ([2409a96](https://github.com/httpland/range-request-middleware/commit/2409a9648e3b68d6e620d75abd261498bf14b2d0))

# [1.0.0-beta.3](https://github.com/httpland/range-request-middleware/compare/1.0.0-beta.2...1.0.0-beta.3) (2023-03-26)


### Bug Fixes


middleware.ts
// Copyright 2023-latest the httpland authors. All rights reserved. MIT license.
// This module is browser compatible.

import {
  ConditionalHeader,
  isNotEmpty,
  isNull,
  Method,
  type Middleware,
  RangeHeader,
} from "./deps.ts";
import { UnitLike, withAcceptRanges, withContentRange } from "./transform.ts";
import { BytesRange } from "./ranges/bytes.ts";
import { RangeUnit } from "./utils.ts";
import type { Range } from "./types.ts";

/** Create range request middleware.
 *
 * @example
 * ```ts
 * import { rangeRequest } from "https://deno.land/x/range_request_middleware@$VERSION/middleware.ts";
 * import {
 *   assert,
 *   assertEquals,
 *   assertThrows,
 * } from "https://deno.land/std/testing/asserts.ts";
 *
 * const middleware = rangeRequest();
 * const request = new Request("test:", {
 *   headers: { range: "bytes=5-9" },
 * });
 * const response = await middleware(
 *   request,
 *   () => new Response("abcdefghijklmnopqrstuvwxyz"),
 * );
 *
 * assertEquals(response.status, 206);
 * assertEquals(response.headers.get("content-range"), "bytes 5-9/26");
 * assertEquals(response.headers.get("accept-ranges"), "bytes");
 * assertEquals(await response.text(), "fghij");
 * ```
 */
export function rangeRequest(
  ranges?: Iterable<Range>,
): Middleware {
  const $ranges = ranges ?? [new BytesRange()];
  const units = Array.from($ranges).map((range) => range.rangeUnit);
  const unitLike = isNotEmpty(units) ? units : RangeUnit.None;

  const contentRangeMiddleware = contentRange($ranges);
  const acceptRangesMiddleware = acceptRanges(unitLike);

  return (request, next) => {
    return contentRangeMiddleware(
      request,
      (request) => acceptRangesMiddleware(request, next),
    );
  };
}

export function contentRange(ranges?: Iterable<Range>): Middleware {
  const $ranges = ranges ?? [new BytesRange()];

  return async (request, next) => {
    const rangeValue = request.headers.get(RangeHeader.Range);

    // A server MUST ignore a Range header field received with a request method that is unrecognized or for which range handling is not defined. For this specification, GET is the only method for which range handling is defined.
    // @see https://www.rfc-editor.org/rfc/rfc9110#section-14.2-4
    if (
      request.method !== Method.Get ||
      isNull(rangeValue) ||
      request.headers.has(ConditionalHeader.IfRange)
    ) return next(request);

    const response = await next(request);

    return withContentRange(response, { ranges: $ranges, rangeValue });
  };
}

export function acceptRanges(unitLike?: UnitLike): Middleware {
  const rangeUnit = unitLike ?? RangeUnit.Bytes;

  return async (request, next) => {
    const response = await next(request);

    return withAcceptRanges(response, rangeUnit);
  };
}
